### ISSUE TITLE
Suggestion: Update from LibGDX 1.7.2 to latest stable or 1.7.3-SNAPSHOT

### ISSUE THREAD

**Issue by Lucrecious** [GoranM/bdx#643]
I noticed that controller connection/disconnection(on ControllerListener) was not implemented properly on the version of LibGDX that Bdx is using (1.7.2), so hotplugging isn't possible (even with polling from Controllers.getControllers()).

If there is an alternate way to achieve hotplugging, please show me. Otherwise, here's a thread with this information I mentioned:
https://www.bountysource.com/issues/3405158-detect-controller-hotplugging

Here's the Controller wiki:
https://github.com/libgdx/libgdx/wiki/Controllers

Thanks.

**Comment by SolarLune** [comment]
Ah, I remember reading about this. I don't think there is a better way to achieve hotplugging (at the moment, at least).

Updating the version of LibGDX might be a good idea to support this.

**Comment by Lucrecious** [comment]
Do you know how we can update it? I can't figure out a way to do it without getting a bunch of errors.

**Comment by SolarLune** [comment]
I meant updating the version in master, like moving BDX to the latest version of LibGDX.

But if you want to update just your own version, it should be possible. Basically, you need to do [what Goran did when we updated BDX to 1.7.2](https://github.com/GoranM/bdx/commit/17953c52cffeb53881018d99218b9ee97a56b246). Looks like he mainly adjusted the build.gradle and build.xml files, and maybe replaced gdx-setup with the latest version from LibGDX. [After that, I believe you might have to create a brand new project in Blender and run that to make LibGDX download the latest dependencies. From there, you might be able to proceed as usual?](https://github.com/GoranM/bdx/issues/235#issuecomment-167465877)

**Comment by GoranM** [comment]
> Looks like he mainly adjusted the build.gradle and build.xml files, and maybe replaced gdx-setup with the latest version from LibGDX.

Basically, but you can't just change the version variables in the existing build.gradle file, because that might miss whatever changes LibGDX people made in the new build.gradle file. So, in reality, you have to replace the old build.gradle file with a new build.gradle file that's generated by the new gdx-setup program. Then you add in the things that BDX needs on top of that (fileTree directives to include the relevant bdx jars, and the gdx-controllers plugin), and that gives you the proper file that can be committed to master.

Do you guys want me to do this, or does someone else want to give it a try?

**Comment by SolarLune** [comment]
I'm good for now, haha; feel free to do it.

**Comment by Lucrecious** [comment]
I would give it a try if I was more familiar with this kind of stuff - I think would take me much longer to do it than it would for you.

If you ever get the chance, it'd be cool if you did it :)

**Comment by GoranM** [comment]
Ok, it's done.

I also refactored project creation, so instead of having our own copy of build.gradle, we're just going to automatically modify the one created by gdx-setup. So, in the future, updating libgdx will just be a matter of replacing gdx-setup.jar, and changing the libgdx version number to match in our build.xml file.

**Comment by Lucrecious** [comment]
This is awesome :D I'm trying it out, looks like I can't just pull and build through ant - I get some errors.

How should I be upgrading?

Thanks.

**Comment by SolarLune** [comment]
I believe this is due to some changes to the API. I'll make a PR to fix it.

**Comment by SolarLune** [comment]
OK, that should do it. PR submitted. Was messing around with some Git stuff; getting used to doing it in IntelliJ, haha.

Anyway, it's been fixed, as LibGDX's graphics window size setting functions changed. The new implementation doesn't seem like it supports changing the resolution of the game when not full-screened, which is a bit of a shame, but I don't think it worked on Linux anyway, so whatever.

**Comment by GoranM** [comment]
Ah, I forgot to do an `ant clean` after updating libgdx, so I didn't see that.

Thanks for the quick fix.

**Comment by SolarLune** [comment]
No problem!

@Lucrecious This should open up the possibility to enable hotplugging, so you should be able to add in hotplugging support to BDX's controllers fairly easily (I think). We're not exposing listeners or callbacks to the developer under BDX, so any reporting of "just connected" or "just disconnected" might be easily done using a variable or function that returns true for a single frame (like `justConnected()` or `justDisconnected()` or something like that).

**Comment by Lucrecious** [comment]
Sounds good! Although, but I'm leaning to write my own wrapper through LibGDX. Partly because BDX uses the custom GamePad wrapper which is weird for multiple joysticks for a single player game. It would be a little bit a of hassle to allow the user to use any controller that's connected to the computer because you'd have to iterate through all the connected GamePads every time you wanted to check for input. Instead, listeners do that for me, and I can set them up to listen for input changes on all controllers at the same time - which is pretty nifty and clean imo.

Thanks again, I'll be testing it out later today, this is awesome.

**Comment by SolarLune** [comment]
> It would be a little bit a of hassle to allow the user to use any controller that's connected to the computer because you'd have to iterate through all the connected GamePads every time you wanted to check for input. 

No, you would just have to iterate through them to determine which gamepad is the active one (or just have the player designate which gamepad they're using). After that, you know which one to check. 

But it's up to you, anyway. If this solved your issue, feel free to close it.

**Comment by Lucrecious** [comment]
Well at any point the player can switch gamepads, so either I need to consistently check which is the active one, or listen for any input coming from any of them. Either way, the solutions are pretty trivial (as in it doesn't really matter which way it's done).

Also, I noticed upgrading my project for this version of BDX isn't as easy as just synchronizing the libs. How can I open my project with the latest version of BDX with the latest version of LibGDX? What files do I need to replace in my project folder?

**Comment by SolarLune** [comment]
> Well at any point the player can switch gamepads, so either I need to consistently check which is the active one, or listen for any input coming from any of them. Either way, the solutions are pretty trivial (as in it doesn't really matter which way it's done).

Maybe if you're developing a game in which the player can switch gamepads for gameplay, but I think most games only check for the active one (as you wouldn't want other gamepads to interrupt the player's gameplay). If the player wants to switch, they rebind controls or change which gamepad index is the active one in the options menu. I don't know of any games that check for input control on multiple controllers during gameplay. But it's up to you; it just sounds weird.

> Also, I noticed upgrading my project for this version of BDX isn't as easy as just synchronizing the libs. How can I open my project with the latest version of BDX with the latest version of LibGDX? What files do I need to replace in my project folder?

Yeah, I just recreated the project folder (so rename the old folder to something else, and create a new project with the same classpath and directory) and copied the relevant files (blend file, assets, code, etc). It's fairly simple to do. 

If you don't want to do that, you'll have to modify some of the files, like the build.gradle file, to point to the new 1.9.2 LibGDX version.

**Comment by Lucrecious** [comment]
> Maybe if you're developing a game in which the player can switch gamepads for gameplay, but I think most games only check for the active one (as you wouldn't want other gamepads to interrupt the player's gameplay). If the player wants to switch, they rebind controls or change which gamepad index is the active one in the options menu. I don't know of any games that check for input control for a single game object on multiple controllers during gameplay. You know what'll work for your project, of course; it just sounds kinda unnecessary.

Yeah, I realized my method was kind of weird after I thought about it for a while, heh.

> Yeah, I just recreated the project folder (so rename the old folder to something else, and create a new project with the same classpath and directory) and copied the relevant files (blend file, assets, code, etc). It's fairly simple to do.
> 
> If you don't want to do that, you'll have to modify some of the files, like the build.gradle file, to point to the new 1.9.2 LibGDX version.

Alright, this is what I thought. Sounds good!

So I've updated my version of LibGDX to 1.9.2. Hotplugging still doesn't work (I'm assuming this is not a problem with the integration with LibGDX but a bug with Controllers itself). I made sure to use a brand new project build to make sure it wasn't just me upgrading my project incorrectly. To check if hotplugging was working, I used:
`Controllers.getControllers().size`
to see if the number changed according to me disconnecting and connecting my controllers.

Is this not the way to get connected controllers? Do I need to refresh something?

**Comment by SolarLune** [comment]
Did you create a new project to download the new dependencies?
_____________

I think controller disconnections would be reported, so the size of the list returned might only change if new controllers are added? I'm not 100% sure on how it works.

**Comment by rafcolson** [comment]
>EDIT: Oh, and make sure you make a new BDX project before you try to build in Ant, @Lucrecious; it'll need to download the new dependencies.

@SolarLune If I just forked current master (so started a new repo), what should I do next to ensure the new dependencies are downloaded?

**Comment by Lucrecious** [comment]
@SolarLune I haven't been notified either way (with connections or without). I'm using the ControllerListener callback on disconnect and connect. I think regardless, the Controllers.getControllers() should get me the complete list of have been, not necessarily still, connected controllers. So if I start with an empty list of have been connected controllers, I should at least get a list of length one when I connect any controller.

**Comment by Lucrecious** [comment]
@GoranM @SolarLune @rafcolson have any of you guys tried hot plugging controllers?

**Comment by SolarLune** [comment]
Hey, sorry. I've tried it, but the controller list doesn't change, like you mentioned. Maybe we're not using the latest version of gdx-controllers? According to [this](https://github.com/libgdx/libgdx/issues/2182#issuecomment-177152147) (kinda old) post on LibGDX's repo, badlogic says that you'd need to use LWJGL3 (which I don't believe we are, currently).

**Comment by GoranM** [comment]
No, and I don't currently have a controller, so I can't really test it.

If there was an update for gdx-controllers, it should have been incorporated with the recent libgdx update.

**Comment by SolarLune** [comment]
Yo, so it seems like for hotplugging, you'd need to switch to the LWJGL3 Controllers and LWJGL3 backend. Then we'd need to change things to remove the old (current) LWJGL references - see the posts that were added in that issue I linked.

**Comment by GoranM** [comment]
As far as I know, "gdx-controllers" is the official cross platform library that should provide all available controller features.

If there's a feature that "gdx-controllers" doesn't provide, but that is available on a specific platform, that feature should be added to "gdx-controllers".
